File: Difficulty: Medium/K Sum Paths/k-sum-paths.cpp
/*
class Node {
  public:
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/
class Solution {
  public:
    void dfs(int &ans , Node *nd , int k , int sum, map<int,int>&mp){
        sum+=nd->data;
        if(mp.find(sum-k) !=mp.end())ans+=mp[sum-k];
        mp[sum]++;
        if(nd->left)dfs(ans , nd->left , k , sum ,mp);
        if(nd->right)dfs(ans , nd->right , k , sum , mp);
        
        if(mp[sum] ==1)mp.erase(sum);
        else mp[sum]--;
    }
    int countAllPaths(Node *root, int k) {
        // code here
        if(root ==NULL)return 0;
        int ans = 0;
        map<int,int>mp;
        mp[0] = 1;
        dfs(ans , root , k , 0,mp );
        return ans;
    }
};

File: Difficulty: Medium/K Sum Paths/k-sum-paths.java
/*
class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public int countAllPaths(Node root, int k) {
        if (root == null) return 0;

        return countFromNode(root, k) 
             + countAllPaths(root.left, k) 
             + countAllPaths(root.right, k);
    }

    int countFromNode(Node node, int k) {
        if (node == null) return 0;

        int count = 0;

        if (node.data == k)
            count++;

        count += countFromNode(node.left, k - node.data);
        count += countFromNode(node.right, k - node.data);

        return count;
    }
}